network/netidentity.go

```diff
@@ -0,0 +1,398 @@
+		Key:           i.identityKeys.SignatureVerifier,
+		Challenge:     newIdentityChallengeValue(),
+		PublicAddress: []byte(addr),
+	}
+
+	attach.Add(IdentityChallengeHeader, c.signAndEncodeB64(i.identityKeys))
+	return c.Challenge
+}
+
+// VerifyAndAttachResponse  headers for an Identity Challenge, and verifies:
+// * the provided challenge bytes matches the one encoded in the header
+// * the identity challenge verifies against the included key
+// * the "Address" field matches what this scheme expects
+// once verified, it will attach the header to the "attach" header
+// and will return the challenge and identity of the peer for recording
+// or returns empty values if the header did not end up getting set
+func (i identityChallengePublicKeyScheme) VerifyAndAttachResponse(attach http.Header, h http.Header) (identityChallengeValue, crypto.PublicKey, error) {
+	// if dedupName is not set, this scheme is not configured to exchange identity
+	if i.dedupName == "" {
+		return identityChallengeValue{}, crypto.PublicKey{}, nil
+	}
+	// if the headerString is not populated, the peer isn't participating in identity exchange
+	headerString := h.Get(IdentityChallengeHeader)
+	if headerString == "" {
+		return identityChallengeValue{}, crypto.PublicKey{}, nil
+	}
+	// decode the header to an identityChallenge
+	msg, err := base64.StdEncoding.DecodeString(headerString)
+	if err != nil {
+		return identityChallengeValue{}, crypto.PublicKey{}, err
+	}
+	idChal := identityChallengeSigned{}
+	err = protocol.Decode(msg, &idChal)
+	if err != nil {
+		return identityChallengeValue{}, crypto.PublicKey{}, err
+	}
+	if !idChal.Verify() {
+		return identityChallengeValue{}, crypto.PublicKey{}, fmt.Errorf("identity challenge incorrectly signed")
+	}
+	// if the address is not meant for this host, return without attaching headers,
+	// but also do not emit an error. This is because if an operator were to incorrectly
+	// specify their dedupName, it could result in inappropriate disconnections from valid peers
+	if string(idChal.Msg.PublicAddress) != i.dedupName {
+		return identityChallengeValue{}, crypto.PublicKey{}, nil
+	}
+	// make the response object, encode it and attach it to the header
+	r := identityChallengeResponse{
+		Key:               i.identityKeys.SignatureVerifier,
+		Challenge:         idChal.Msg.Challenge,
+		ResponseChallenge: newIdentityChallengeValue(),
+	}
+	attach.Add(IdentityChallengeHeader, r.signAndEncodeB64(i.identityKeys))
+	return r.ResponseChallenge, idChal.Msg.Key, nil
+}
+
+// VerifyResponse will decode the identity challenge header and confirm it self-verifies,
+// and that the provided challenge matches the encoded one
+// if the response can be verified, it returns the identity of the peer and a final Verification Message to send to the peer
+// otherwise, returns empty values
+func (i identityChallengePublicKeyScheme) VerifyResponse(h http.Header, c identityChallengeValue) (crypto.PublicKey, []byte, error) {
```

Comment:
The only way for you to receive a ChallengeResponse to verify is if you:
1) Initiate the connection
2) Sent an Identity Challenge in the initial request

If there is an attached Challenge Response but the requester wasn't expecting one, that indeed sounds like an out-of-protocol behavior?

Attaching a dedupName check would avoid the behavior you're describing, but that behavior is already an unexpected one IMO.

Reply:
> shouldn't you ignore X-Algorand-IdentityChallenge headers always

That is a valid way to handle it, sure, and if we'd like to change it to incorporate that behavior, I am totally fine with that.

But, if a node sends a non-identified request, and receives a response with an identity challenge response, there is no situation where that's legitimate. Since we know we never sent the peer an original challenge, we therefore also know this challenge response couldn't possibly be valid. Therefore, we know the peer is sending a response with inappropriate headers.

It's not just that the requester node is not participating in identity. It's that the peer being connected to *should not* respond with an identity challenge response. The correct response to a peering request without identity challenge is a peering response *also without identity challenge*.

The difference in behavior here is small, and like I said I'm fine to adopt whichever you like:

* As-written: if a node ever sends and identity challenge response when there is none expected, it will inevitably fail verification and the connection will be dropped. This means that if a responding peer should ever deviate from protocol (attach a challenge response when there was no original), it is considered worthy of disconnect.

* As-suggested: if a node ever sends and identity challenge response when there is none expected, it will be ignored and the peering will continue without identity. The deviation from protocol is acceptable in this design.

Are there situations you're thinking of where this can happen with two protocol-conforming peers?

Reply:
sure, I have made this change.